src/Servers/Kestrel/Core/src/Middleware/HttpsConnectionMiddleware.cs

```diff
@@ -215,6 +226,52 @@ await using (sslDuplexPipe)
             }
         }
 
+        // This logic is replicated from https://github.com/dotnet/runtime/blob/02b24db7cada5d5806c5cc513e61e44fb2a41944/src/libraries/System.Net.Security/src/System/Net/Security/SecureChannel.cs#L195-L262
+        // but with public APIs
+        private X509Certificate2 LocateCertificateWithPrivateKey(X509Certificate2 certificate)
+        {
+            Debug.Assert(!certificate.HasPrivateKey, "This should only be called with certificates that don't have a private key");
+
+            _logger.LocatingCertWithPrivateKey(certificate);
+
+            try
+            {
+                using (var store = new X509Store(StoreName.My, StoreLocation.LocalMachine))
+                {
+                    store.Open(OpenFlags.ReadOnly);
+
+                    var certs = store.Certificates.Find(X509FindType.FindByThumbprint, certificate.Thumbprint, validOnly: false);
+
+                    if (certs.Count > 0 && certs[0].HasPrivateKey)
+                    {
+                        _logger.FoundCertWithPrivateKey(certs[0], StoreLocation.LocalMachine);
+                        return certs[0];
+                    }
+                }
+
+                using (var store = new X509Store(StoreName.My, StoreLocation.CurrentUser))
```

Comment:
I would put this to separate try/catch. If opening StoreLocation.LocalMachine fails for what ever reason, we would not get here.

Reply:
That's how the current code in the BCL is written though right?

Reply:
not quite. 
```c#
 X509Store? store = CertificateValidationPal.EnsureStoreOpened(isServer);
```
will eat exception and return `null` on failure.